### ISSUE TITLE
[FEATURE] SR-IOV Support

### ISSUE THREAD

**Issue by tlehman** [harvester/harvester#2763]
**Is your feature request related to a problem? Please describe.**
A customer could not use their 10Gb NIC to it's full potential inside a VM, this prompted #992, but that ticket is now dedicated to tracking PCI Passthrough, not SR-IOV specifically.

**Describe the solution you'd like**
The ability to create VFs for SR-IOV-capable devices, then pass them through to VMs. 

**Describe alternatives you've considered**
- [sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator), this would not be a good fit for our customer's needs. In the host cluster it would use CNI and not allow for hot-adding of VFs to the VM. In the guest cluster it would constrain our users to 1 PF per VM, which is too limiting.

**Additional context**
This issue (#992) has lead to a lot of study, both of technical requirements and business requirements. While this is a feature that we are going to build, the PCI Passthrough feature is more generic and flexible, and is a dependency of this ticket. This is why we prioritized it and pushed the rest of the SR-IOV functionality into this new ticket.

**Comment by wjimenez5271** [comment]
Can we also test the following NIC cards for compatibility with SR-IOV in Harvester? 
- Intel 520 
- ~~Intel X710~~
- Mellanox ConnectX-5

**Comment by irishgordo** [comment]
Still testing but noticed https://github.com/harvester/harvester/issues/3977 - so opened that up and drafted a small pr on dashboard that 'might' fix it but tbd if that encompasses what's needed for the fix

**Comment by irishgordo** [comment]
**validation issue**
@ibrokethecloud  - I'm running into an issue with the Mellanox ConnectX-5 where:
```yaml
  state:
    error: true
    message: >-
      0/3 nodes are available: 3 Insufficient
      mellanox.com/MT27800_FAMILY_CONNECTX5_VIRTUAL_FUNCTION. preemption: 0/3
      nodes are available: 3 No preemption victims found for incoming pod.
    name: scheduling
    transitioning: false
  uid: d5511be8-f257-4605-9659-8a1d495ef295
```
The VM with the virtual function attached never seems to start.
On the node the `ip link` shows:
```
4: ens1f0np0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether 24:8a:07:a5:29:88 brd ff:ff:ff:ff:ff:ff
    vf 0     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 1     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 2     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 3     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 4     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 5     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 6     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 7     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    altname enp8s0f0np0
5: ens1f1np1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether 24:8a:07:a5:29:89 brd ff:ff:ff:ff:ff:ff
    vf 0     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 1     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 2     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 3     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 4     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 5     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 6     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off
    vf 7     link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff, spoof checking off, link-state auto, trust off, query_rss off

```
Referencing:
```
  *-network:0
       description: Ethernet interface
       product: MT28800 Family [ConnectX-5 Ex]
       vendor: Mellanox Technologies
       physical id: 0
       bus info: pci@0000:08:00.0
       logical name: ens1f0np0
       version: 00
       serial: 24:8a:07:a5:29:88
       capacity: 40Gbit/s
       width: 64 bits
       clock: 33MHz
       capabilities: pciexpress vpd msix pm bus_master cap_list rom ethernet physical fibre 1000bx-fd 10000bx-fd 25000bx-fd 40000bx-fd autonegotiation
       configuration: autonegotiation=on broadcast=yes driver=mlx5_core driverversion=5.14.21-150400.24.63-default duplex=full firmware=16.18.0160 (MT_0000000013) latency=0 link=yes multicast=yes
       resources: irq:16 memory:94000000-95ffffff memory:39ff7c00000-39fff9fffff
  *-network:1
       description: Ethernet interface
       product: MT28800 Family [ConnectX-5 Ex]
       vendor: Mellanox Technologies
       physical id: 0.1
       bus info: pci@0000:08:00.1
       logical name: ens1f1np1
       version: 00
       serial: 24:8a:07:a5:29:89
       capacity: 40Gbit/s
       width: 64 bits
       clock: 33MHz
       capabilities: pciexpress vpd msix pm bus_master cap_list rom ethernet physical fibre 1000bx-fd 10000bx-fd 25000bx-fd 40000bx-fd autonegotiation
       configuration: autonegotiation=on broadcast=yes driver=mlx5_core driverversion=5.14.21-150400.24.63-default firmware=16.18.0160 (MT_0000000013) latency=0 link=no multicast=yes port=fibre
       resources: irq:17 memory:96000000-97ffffff memory:39fefe00000-39ff7bfffff
```
[supportbundle_07b89afa-696a-430f-9b93-9808a2a6e499_2023-05-30T22-15-10Z.zip](https://github.com/harvester/harvester/files/11606483/supportbundle_07b89afa-696a-430f-9b93-9808a2a6e499_2023-05-30T22-15-10Z.zip)
![Screenshot from 2023-05-30 15-14-27](https://github.com/harvester/harvester/assets/5370752/c14c1213-d7cc-4998-8f1c-ba8b4129a82a)

**Comment by irishgordo** [comment]
@ibrokethecloud 

With an [Intel 520/X520](https://ark.intel.com/content/www/us/en/ark/products/series/32609/intel-82599-10-gigabit-ethernet-controller.html) series based PCI card, things look good locally being able to create the VFs on the card, have them attached to a VM, allow for the VM to have traffic routing through them, also setting up VLANs via the VF on the VM that comes from the physical device - tested locally with a X520 card, MicroTik CRS305-1G-4S+IN(Switch/Router/Bridge) that I recently purchased to connect to my existing Ubiquiti Unifi Gear, in Bridge Mode.

https://github.com/harvester/harvester/assets/5370752/610eb266-3e76-4e09-b4c4-e93aa5e36cbe

On v1.2.0-rc1 :smile: :+1:

**Comment by irishgordo** [comment]
@ibrokethecloud 
**validation issue**

I've noticed some strange behavior with errors related to maintenance mode and running vms with sr-iov vfs, that are present on a node that is placed in maintenance mode, I've opened up:
https://github.com/harvester/harvester/issues/4015

And also:
https://github.com/harvester/harvester/issues/4017

**Comment by irishgordo** [comment]
@ibrokethecloud after checking 4015, things look great on Harvester Version: 2dc1dc0d-dirty :smile:  - 

Should I leave this open until #4017 is closed or should I go ahead and close this out?

**Comment by ibrokethecloud** [comment]
@irishgordo technically the underlying PF can be used, which is the point of the virtualization.

I don't feel there is a plan to stop this behavior at the moment.

**Comment by irishgordo** [comment]
@ibrokethecloud I've gone ahead and closed out #4017 as that is expected behavior :smile:  
So that said I'll go ahead and close this out as everything looks great :smile: :+1:
